src/Linters/DontShadowLocalVariablesInForeachLinter.hack

```diff
@@ -0,0 +1,164 @@
+/*
+ *  Copyright (c) 2017-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the MIT license found in the
+ *  LICENSE file in the root directory of this source tree.
+ *
+ */
+
+namespace Facebook\HHAST;
+
+use namespace HH\Lib\{C, Dict, Keyset, Str, Vec};
+
+final class DontShadowLocalVariablesInForeachLinter extends ASTLinter {
+  const type TContext = IFunctionishDeclaration;
+  const type TNode = ForeachStatement;
+
+  <<__Override>>
+  public function getLintErrorForNode(
+    this::TContext $scope,
+    this::TNode $foreach,
+  ): ?ASTLintError {
+    $pseudo_declarations = static::getPseudoDeclarations($foreach);
+    if (C\is_empty($pseudo_declarations)) {
+      return null;
+    }
+    $scope_decl = i_functionish_get_decl_header($scope);
+    $params = $scope_decl->getParameterList()
+      ?->getDescendantsOfType(VariableToken::class)
```

Comment:
This can be incorrect if any of the parameters has a lambda function as its default value (I recently had to fix a similar bug in https://github.com/hhvm/hhast/pull/342, you can look at the test cases there -- maybe copy some of them here too).

It should probably only look at direct children, not all descendants.

You might also be able to extract and reuse some code from UnusedVariableLinter, e.g. the parts where it computes the list of params from a lambda/non-lambda function (getLambdaParts, getFunctionishParts).